hcl/eval/partial.go

```diff
@@ -54,30 +54,59 @@ Here be dragons. Thou art forewarned
            `B'\)                  ///,-'~`__--^-  |-------~~~~^'
            /^>                           ///,--~`-\
           `  `
+
+
+This implementation is based on HCL v2.11.1 as defined in the "spec" below:
+  - https://github.com/hashicorp/hcl/blob/v2.11.1/hclsyntax/spec.md
+
+Motivation:
+
+Terramate supports a HCL code generation feature driven by pure HCL, where
+everything with the exception of terramate variables must be copied verbatim
+into the generated file. This code is needed because the Hashicorp's hcl library
+does not support the partial evaluation of expressions, in other words, given an
+expression containing unknown symbols and terramate symbols, the hcl
+expression.Value(ctx) fails because all symbols must be populated in the context
+and this is obviously not possible for generating arbitrary code.
+
+The solution here involves lexing/scanning the expression bytes into tokens,
+then parsing them and evaluating the terramate symbols as they are found,
+rewriting the token stream with the primitive values, and keeping everything
+else untouched.
 */
 
-type Engine struct {
-	tokens    hclwrite.Tokens
-	pos       int
-	ctx       *Context
-	evaluated []hclwrite.Tokens
+type engine struct {
+	tokens hclwrite.Tokens
+	pos    int
+	ctx    *Context
+
+	// evalstack is a stack of evaluated nodes.
+	// The engine walks through the token list evaluating them as needed into a
+	// separated node struct placed in this stack.
+	evalstack nodestack
 
 	nparen int
 }
 
-func NewEngine(tokens hclwrite.Tokens, ctx *Context) *Engine {
-	return &Engine{
+// node represents a grammar node but in terms of its original source tokens and
+// the rewritten (evaluated) ones.
+type node struct {
+	source    hclwrite.Tokens
+	evaluated hclwrite.Tokens
```

Comment:
I think `original` is ok but the name `tokens` is confusing if it's about the result tokens or input tokens, maybe but I'd have to change the `e.tokens` to `e.input` but even though not sure. About the `evaluated`, in my mind is ok because eval doesn't change the value all times, for example: `eval("1") == "1"`, so it's ok to be "evaluated" but keep the same value.